tests/asyncio_test_helpers.py

```diff
@@ -0,0 +1,162 @@
+import logging
+import types
+from asyncio import AbstractEventLoop, transports
+from asyncio.protocols import BaseProtocol, Protocol
+from asyncio.transports import Transport
+from typing import Any, List, Tuple
+
+logger = logging.getLogger(__name__)
+
+
+class TimelessEventLoopWrapper:
+    @property  # type: ignore
+    # -- we have no need to make this read/write
+    # (o.O being able to re-write a __class__ is odd!)
+    def __class__(self):
+        """
+        Fakes isinstance(this, AbstractEventLoop) so we can set_event_loop
+         without fail.
+        """
+        return self._wrapped_loop.__class__
+
+    def __init__(self, wrapped_loop: AbstractEventLoop):
+        self._wrapped_loop = wrapped_loop
+        self._time = 0.0
+        self._to_be_called: List[Tuple[float, Any, Any, Any]] = []
+
+    def advance(self, time_delta: float):
+        target_time = self._time + time_delta
+        logger.debug(
+            "advancing from %f by %f (%d in queue)",
+            self._time,
+            time_delta,
+            len(self._to_be_called),
+        )
+        while self._time < target_time and self._to_be_called:
+            # pop off the next callback from the queue
+            next_time, next_callback, args, _context = self._to_be_called[0]
+            if next_time > target_time:
+                # this isn't allowed to run yet
+                break
+            logger.debug("callback at %f on %r", next_time, next_callback)
+            self._to_be_called = self._to_be_called[1:]
+            self._time = next_time
+            next_callback(*args)
+
+        # no more tasks can run now but advance to the time anyway
+        self._time = target_time
+
+    def __getattr__(self, item):
+        """
+        We use this to delegate other method calls to the real EventLoop.
+        """
+        value = getattr(self._wrapped_loop, item)
+        if isinstance(value, types.MethodType):
+            # rebind this method to be called on us
+            # this makes the wrapped class use our overridden methods when
+            # available.
+            # Otherwise, call_soon etc won't come to us in all scenarios.
+            return types.MethodType(value.__func__, self)
+        else:
+            return value
+
+    def call_later(self, delay, callback, *args, context=None):
+        self.call_at(self._time + delay, callback, *args)
+
+    def call_at(self, when, callback, *args, context=None):
+        logger.debug(f"Calling {callback} at %f...", when)
+        self._to_be_called.append((when, callback, args, context))
+
+        # re-sort list in ascending time order
+        self._to_be_called.sort(key=lambda x: x[0])
+
+    def call_soon(self, callback, *args, context=None):
+        return self.call_later(0, callback, *args, context=context)
+
+    def time(self):
+        return self._time
+
+
+class MockTransport(Transport):
```

Comment:
The Java programmer in me felt that this should be.
But you're right, the duck typingness of Python means we need not explicitly override Transport ourselves.

Reply:
Oh:
```
builtins.ValueError: transport must be a proper Transport
```

I think Twisted devs also think in a Java way, a lot of the Twisted library gives off Java vibes anyway :). I don't think we can do anything clean about this, regardless of what is best.

Reply:
Note that `Transport` is an interface (again, a Javaism).

I think it's somewhat clear that they'd consider changing the interface a breaking change in Twisted, much like it would in Java.

Reply:
Isn't this all asyncio, not Twisted?

Reply:
…. urr, yes. I suppose the similarity has finally gotten to me and confused me. Oops.